### PR TITLE
Question-Answer Transformation using a Quora Question Paraphraser

### PR DESCRIPTION
**Pull request by kaustubhdhole**
1. This transformation creates new QA pairs by generating question paraphrases from a T5 model fine-tuned on Quora Question pairs.   
2. This transformation acts like a paraphrase generator. Generated questions can have a very different surface form from the original question making it a strong paraphrase generator. A T5 model was fine tuned on the Quora Question Pairs dataset and is being used to generate question paraphrases.
3. Here is the performance of the model mrm8488/bert-tiny-5-finetuned-squadv2 on the validation[:20%] split of the squad dataset
The accuracy on a subset of squad = 60.31220435193945
The accuracy on its perturbed set generated from QuoraT5QaPairGenerator = 47.161778618732264
This ~13% would also be attributed to paraphrasing errors.

### CODE REVIEW

File: transformations/quora_trained_t5_for_qa/README.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Quora Trained T5 for question paraphrases for QA Task ❓️ → 🐍❔
+This transformation creates new QA pairs by generating question paraphrases from a T5 model fine-tuned on Quora Question pairs.
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/README.md:2]: #BenderRule

File: transformations/quora_trained_t5_for_qa/__init__.py [hunk 2]
```diff
@@ -0,0 +1 @@
+from .transformation import *
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/__init__.py:1]: nit: new newline at the end of the file.

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 3]
```diff
@@ -0,0 +1,55 @@
+import torch
+from transformers import T5ForConditionalGeneration, T5Tokenizer
+
+from interfaces.QuestionAnswerOperation import QuestionAnswerOperation
+from tasks.TaskTypes import TaskType
+import random
+
+"""The T5 model and code has been taken from https://huggingface.co/ramsrigouthamg/t5_paraphraser. The test cases
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:8]: This comment reads a bit strange. Maybe follow the convention to write one sentence high-level before adding details.

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+from tasks.TaskTypes import TaskType
+import random
+
+"""The T5 model and code has been taken from https://huggingface.co/ramsrigouthamg/t5_paraphraser. The test cases 
+were generated by the commented script in redundant_context_for_qa. 
+
+"""
+
+
+class QuoraT5QaPairGenerator(QuestionAnswerOperation):
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:14]: Doctrings please :)

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    ]
+    languages = ["en"]
+    heavy = True
+
+    def __init__(self, seed=0, model_name="ramsrigouthamg/t5_paraphraser", max_len=256):
+        super().__init__(seed)
+        self.model = T5ForConditionalGeneration.from_pretrained(model_name)
+        self.tokenizer = T5Tokenizer.from_pretrained(model_name)
+        self.max_len = max_len
+        self.num_return_sequences = 3
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:27]: Making this configurable could help.

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        random.seed(self.seed)
+        if torch.cuda.is_available():
+            torch.cuda.manual_seed_all(self.seed)
+
+    def generate(self, context: str, question: str, answers: [str]):
+
+        text = "paraphrase: " + question + "</s>"
+        encoding = self.tokenizer.encode_plus(text, padding="longest", return_tensors="pt")
+        input_ids, attention_masks = encoding["input_ids"], encoding["attention_mask"]
+        beam_outputs = self.model.generate(
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:38]: It could be a helpful addition to add a filtering step of some sort for cases where the outputs are of insufficient quality.

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+            input_ids=input_ids, attention_mask=attention_masks,
+            do_sample=True,
+            max_length=256,
+            top_k=120,
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:42]: 120 is quite a bit, maybe reducing to 40 would lead to better quality (albeit less diversity.)

File: transformations/quora_trained_t5_for_qa/transformation.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+            top_p=0.98,
+            early_stopping=True,
+            num_return_sequences=self.num_return_sequences
+        )
+        unique_question_paraprases = []
+        for beam_output in beam_outputs:
+            sent = self.tokenizer.decode(beam_output, skip_special_tokens=True, clean_up_tokenization_spaces=True)
+            if sent.strip() and sent.lower() != question.lower():
+                unique_question_paraprases.append(sent)
```
Comment (sebastianGehrmann) [inline transformations/quora_trained_t5_for_qa/transformation.py:51]: maybe rename to clarify that this ensures that the paraphrase != original instead of `unique`?

**Review by sebastianGehrmann — CHANGES_REQUESTED** [review]

**Comment by kaustubhdhole** [comment]
Thanks @sebastianGehrmann very much for your comments. I've fixed those.
